Hw24/Vector2d.cpp

```diff
@@ -0,0 +1,147 @@
+#include "Vector2d.h"
+
+Vector2d::Vector2d(const float x, const float y):
+m_x1{x},
+m_y1{y}
+{
+}
+
+Vector2d::Vector2d(const float x1, const float y1, const float x2, const float y2):
+m_x1{ x1 },
+m_y1{ y1 },
+m_x2{ x2 },
+m_y2{ y2 }
+{
+}
+
+Vector2d& Vector2d::operator=(const Vector2d& v)
+{
+	if (this == &v)
+		return *this;
+
+	m_x1 = v.m_x1;
+	m_y1 = v.m_y1;
+
+	return *this;
+}
+
+Vector2d Vector2d::operator+(const Vector2d& secondVector) const
+{
+	Vector2d v(*this);
+
+	v.m_x1 += secondVector.m_x1;
+	v.m_y1 += secondVector.m_y1;
+
+	return v;
+}
+
+Vector2d Vector2d::operator-(const Vector2d& secondVector) const
+{
+	Vector2d v(*this);
+
+	v.m_x1 -= secondVector.m_x1;
+	v.m_y1 -= secondVector.m_y1;
+
+	return v;
+}
+
+Vector2d operator+(const Vector2d& leftVector, const Vector2d& rightVector)
+{
+	Vector2d leftVector1(leftVector);
+	return leftVector1 += rightVector;
+}
+
+Vector2d operator-(const Vector2d& leftVector, const Vector2d& rightVector)
+{
+	Vector2d leftVector1(leftVector);
+	return leftVector1 -= rightVector;
+}
+
+Vector2d& Vector2d::operator+=(const Vector2d& rightVector)
+{
+	m_x1 += rightVector.m_x1;
+	m_y1 += rightVector.m_y1;
+
+	return *this;
+}
+
+Vector2d& Vector2d::operator-=(const Vector2d& rightVector)
+{
+	m_x1 -= rightVector.m_x1;
+	m_y1 -= rightVector.m_y1;
+
+	return *this;
+}
+
+float Vector2d::operator()() const
+{
+	const float res = static_cast<float>(std::sqrt(std::pow(m_x1, 2) + std::pow(m_y1, 2)));
+	return res;
+}
+
+float& Vector2d::operator[](const size_t i)
+{
+	return (&m_x1)[i];
```

Comment:
severe dependency on inner structure of the private members.
Simple operation as swapping m_x1, m_y1 places will break it

Reply:
How swap could break it? can you bring an example?

Reply:
even /fsanitize options are set in the command line, and it is silent

Reply:
So, changing declaration to 
```
private: 
float m_y1;
float m_x1;
```
Will work as expected?

Reply:
Definetly it will break it.
I thought you meant std::swap.
Yeap, this approach is... disaster, if smth's gone wrong, like index  out bound, changing declaration, vptr, etc.

Reply:
Still, this cool way, especially if we would have 4 members.